modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+                         bsa.SpecimenAttributeID as attributeId,
+                         bsa.Label as name,
+                         bsa.DatatypeID as datatypeId,
+                         bsa.ReferenceTableID as refTableId,
+                         bsma.Required as required
+                  FROM biobank_specimen_method_attribute_rel bsma
+                  JOIN biobank_specimen_method bsm
+                    USING (SpecimenMethodID)
+                  JOIN biobank_specimen_attribute bsa
+                    USING (SpecimenAttributeID)";
+        $result = $this->db->pselect($query, array());
+        $mA     = array(); //methodAttributes
+        foreach ($result as $row) {
+            foreach ($row as $value) {
+                $mA[$row['methodId']][$row['attributeId']]['name']       = $row['name'];
+                $mA[$row['methodId']][$row['attributeId']]['datatypeId'] = $row['datatypeId'];
+                $mA[$row['methodId']][$row['attributeId']]['refTableId'] = $row['refTableId'];
+                $mA[$row['methodId']][$row['attributeId']]['required']   = $row['required'];
+            }
+        }
+
+        return $mA;
+    }
+
+    /**
+     * This function receives a Specimen Object, converts it into a SQL format
+     * and inserts it into all relevant tables in the database.
+     *
+     * @param object $specimen The Specimen Object to be saved the
+     *                         database.
+     *
+     * @return void
+     */
+    public function saveInstance(Specimen $specimen) : void
+    {
+        // Converts Specimen Object into SQL format.
+        $specimenArray = $this->getArrayFromInstance($specimen);
+        $data          = $this->getSQLFromArray($specimenArray);
+
+        // TODO: Look into implementing a transaction here
+        // $this->db->_PDO->beginTransaction();
+        // $this->db->_PDO->commit();
+        // $this->db->_PDO->rollBack();
+
+        // Inserts or Updates the biobank_specimen table with respective data.
+        $this->db->insertOnDuplicateUpdate(
+            'biobank_specimen',
+            $data['biobank_specimen']
+        );
+
+        $specimenId = $specimen->getId();
+        if (isset($specimenId)) {
+            // If update:
+            // XXX: Logic will likely eventually go here.
+        } else {
+            // If insert:
+            $specimenId = $this->db->getLastInsertId();
+            // set the respective data array index to the value of the new
+            // specimen ID
+            $parentSpecimenIds = $newSpecimen->getParentSpecimenIds();
+            foreach ($parentSpecimenIds as $id) {
+                $data['biobank_specimen_parent'][$id]
+                    = array(
+                       'SpecimenID'       => $specimenId,
+                       'ParentSpecimenID' => $id,
+                      );
+            }
+            $data['biobank_specimen_freezethaw']['SpecimenID']  = $specimenId;
```

Comment:
So the way I see it, freeze/thaw should be a property of specimens because freeze/thaw is only every tracked on primary containers. Temperature is a property of both primary and nonprimary containers (test-tube, matrix boxes, fridges, etc.) Freeze/thaw is specific to certain specimen types and would never be applied to nonprimary containers such as matrix boxes and fridges.

Reply:
noted